src/libmongoc/src/mongoc/mongoc-aggregate.c

```diff
@@ -92,53 +93,81 @@ _make_agg_cmd (const char *ns,
+   const char *error_hint = "";
+
+   bsonBuild (
+      *command,
+      kv ("aggregate",
+          if (dot,
+              /* Note: we're not validating that the collection name's length is
+                 one or more characters, as functions such as
+                 mongoc_client_get_collection also do not validate. */
+              // If 'ns' contains a dot, insert the string after the dot:
+              then (cstr (dot + 1)),
+              // Otherwise just an integer 1:
+              else(int32 (1)))));
+   if ((error_hint = "append-aggregate", error = bsonBuildError)) {
+      goto fail;
    }
 
    /*
     * The following will allow @pipeline to be either an array of
     * items for the pipeline, or {"pipeline": [...]}.
     */
-   if (bson_iter_init_find (&iter, pipeline, "pipeline") &&
-       BSON_ITER_HOLDS_ARRAY (&iter)) {
-      bson_iter_recurse (&iter, &has_write_key_iter);
-      if (!bson_append_iter (command, "pipeline", 8, &iter)) {
-         bson_set_error (err,
-                         MONGOC_ERROR_COMMAND,
-                         MONGOC_ERROR_COMMAND_INVALID_ARG,
-                         "Failed to append \"pipeline\" to create command.");
-         return false;
-      }
-   } else {
-      BSON_APPEND_ARRAY (command, "pipeline", pipeline);
-      bson_iter_init (&has_write_key_iter, pipeline);
+   bsonParse (
+      *pipeline,
+      find (keyWithType ("pipeline", array),
+            // There is a "pipeline" array in the document
+            append (*command, kv ("pipeline", iterValue (bsonVisitIter)))),
+      else( // We did not find a "pipeline" array. copy the pipeline as
+            // an array into the command
+         append (*command, kv ("pipeline", array (insert (*pipeline, true))))));
+   if ((error_hint = "append-pipeline", error = bsonParseError)) {
+      goto fail;
+   }
+
+   // Check if there is a $merge or $out in the pipeline for the command
+   bool has_write_key = false;
+   bsonParse (*command,
+              find (
+                 // Find the "pipeline" array
+                 keyWithType ("pipeline", array),
+                 parse (
+                    // Find the last element of the pipeline array
+                    find (lastElement,
+                          // If it has an "$out" or "$merge" key, it is a
+                          // writing aggregate command.
+                          parse (find (key ("$out", "$merge"),
+                                       do(has_write_key = true)))))));
+   if ((error_hint = "parse-pipeline", error = bsonParseError)) {
+      goto fail;
    }
 
-   has_write_key = _has_write_key (&has_write_key_iter);
-   bson_append_document_begin (command, "cursor", 6, &child);
-   /* Ignore batchSize=0 for aggregates with $out or $merge */
-   if (opts->batchSize_is_set && !(has_write_key && opts->batchSize == 0)) {
-      BSON_APPEND_INT32 (&child, "batchSize", opts->batchSize);
+   bsonBuildAppend (
+      *command,
+      kv ("cursor",
```

Comment:
```suggestion
      kvl ("cursor",
           6,
```
Suggest preserving length to avoid call to `strlen`.

Reply:
Resolving in response to https://github.com/mongodb/mongo-c-driver/pull/1094#discussion_r1004799519.